clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java

```diff
@@ -618,7 +618,7 @@ public ProducerConfig(Map<String, Object> props) {
         super(CONFIG, props);
     }
 
-    ProducerConfig(Map<?, ?> props, boolean doLog) {
+    protected ProducerConfig(Map<?, ?> props, boolean doLog) {
```

Comment:
Changing the visibility to `protected` is actually a change in the public API since it changes the javadocs (https://kafka.apache.org/36/javadoc/org/apache/kafka/clients/consumer/ConsumerConfig.html) which are the source of truth for the public API. However, since it is a small addition (it want break backwards compatibility), I am fine with it.

Reply:
We did a similar change w/o a KIP in the past (https://github.com/apache/kafka/commit/487b95454233e981a65c184c7f3b86adf34058ef)

I agree it's borderline, but believe it's ok. -- In the end, I would rather exclude `protected` stuff from the JavaDocs as I don't think `protected` is actually public API?

Reply:
protected is public API for classes where inheritance is allowed.

Reply:
Pointing to a commit where you did something doesn't quite provide evidence that it's following the process. ;)

Reply:
Interesting, I hadn't heard this definition...I suppose it makes sense in general, but I would actually argue that this specific scenario demonstrates an edge case where that definition breaks down. Wouldn't this also mean that we cannot even add new private constructors to a public class, just because they would show up in the javadocs? That doesn't sound right...I can see how one could argue that package-private and protected methods are considered part of the public API and should not be changed in a backwards incompatible way. But private methods have javadocs that would change, yet they are definitely not part of the public API. 

(Just curious about this, since I've often wondered if we even have a specific policy definitely what is/isn't strictly speaking considered a public API.)

Reply:
Private methods are not part of the API.

Reply:
> protected is public API for classes where inheritance is allowed.

While technically correct, I don't think that `ProducerConfig` is a class that is _intended_ to be extended (and we cannot make it final because we want to extend it internally). -- As a matter of fact, `ProducerConfig` should not even be instantiated by a user, but it's only public to give access to the config definitions which are all `public static final String ...` -- all other methods it has (`configName()` etc) are also just public for practical reason and should actually be hidden from the user -- it's a very leaky abstraction from an API POV -- we actually have a similar issue for `StreamsConfig`; even worse for this case; for which I consider doing a KIP to clean it up: https://github.com/apache/kafka/pull/14548)

> Pointing to a commit where you did something doesn't quite provide evidence that it's following the process. ;)

Sure, but I wanted to point out that many people did consider this ok in the past, so why do we think it was not ok...?

If you are not going to revert this change if we merge it, I would just move forward with this PR as-is, not doing a KIP (as it's highly questionable)... If you insist on a KIP, Sophie please just write one and skip the discussion so we can VOTE it directly -- but it seems totally unnecessary work to me)

Reply:
@ijuma I can do a very quick KIP if you would like (and kick off the voting immediately as I think the change itself is uncontroversial). Obviously I would prefer not to, but am happy with whatever gets this through the fastest, even if that means just doing a KIP to avoid a long drawn-out debate

Also, just to take a quick step back, we are not saying that "a change to javadocs of a public class" is what defines the public API, but rather anything that changes the javadocs of what is established as a public API then needs a KIP? In other words, the question is not really about the javadocs, but simply whether `protected` APIs are part of the public API? (I hadn't refreshed the page and so didn't see any of the responses after Bruno's initial comment when I wrote my reply above, but I still would like to clarify this concretely)

Reply:
Personally I suppose I can see the argument that `protected` APIs in a public and non-final class are themselves part of the public API, as one is technically free to extend that class and would then need to (or at least be able to) leverage those APIs. 

However...I think one could also argue that the public API is only what is directly accessible to the user, and that if you need to extend a public class in order to access a given API, then that API is not strictly speaking part of the API as it is only indirectly accessible from a custom user-defined class. Imo this is more of a hack/breach of internal APIs than a legitimate use of the public API -- as a user, I would not expect my custom implementation to have any guarantees or to be covered by the public contract, and would considered it my own responsibility/fault if my code failed to compile after an upgrade because there was a change to the `protected` API I was sneakily using via inheritance. 

Just my own two cents about the situation: that frankly, accessing protected classes via inheritance is closer in spirit to accessing them (or even private methods) via reflection, than it is to accessing a true public API covered by the public contract. Just because one can leverage a feature of Java to gain access to an API, that doesn't make it a public API in itself.

All that said, I am still happy to do a KIP if that will be the path of least resistance. @ijuma just let me know because I don't want to merge something that you would feel the need to revert 🙂

Reply:
The KIP page states:

> Any change that impacts the public interfaces of the project

Then it lists examples:

> Any class for which the build generates Javadoc

Hopefully these parts are not controversial. Then the next question is which parts of a class are part of the public interface. This is defined by Java accessibility rules combined with project specific conventions.

When it comes to Java accessibility rules, public classes and methods are obvious. The next bucket are protected methods of non-final public classes - those are also part of the public interface since one can call such methods from a subclass. Newer Java versions also have the concept of sealed classes, but we can ignore those for now.

Finally, you have project-specific conventions. Some projects state explicitly which classes are designed for inheritance and all other ones are not supposed to be inherited (and hence their protected methods are not part of the public interface). Kafka (as far as I know) hasn't done this yet, but we probably should. One example of a project specific convention we have is the InterfaceStability annotation (it relaxes compatibility guarantees, but doesn't affect what's considered part of the public interface).

Going back to the original question, does this need a KIP? If we are to follow the rules of the project, absolutely. If we want to change the rules, we should have the discussion and update the KIP documentation. Otherwise, it's difficult to be fair and consistent.

Reply:
I would be in strong favor of clarifying (or changing) the rules. Requiring a KIP for a change like this seems to defeat he purpose of the KIP process.

Sophie could do a quick KIP in parallel to unblock this PR if she want, but to avoid such discussion in the future and to make it simpler to just "do stuff" that is clearly straightforward, changing/clarifying the rules seems to be worth the effort.

@ableegoldman do you want to drive a discussion to clarify the rules?

Reply:
Fair enough -- I will write up a quick KIP. Thanks for clarifying about the project-specific conventions, or rather, our lack of them. I suppose it would be a good idea to have a discussion around the project conventions and define some concrete rules and/or exceptions, but I will leave that to someone/sometime else

Reply:
Here is the KIP for this change: https://cwiki.apache.org/confluence/display/KAFKA/KIP-998%3A+Give+ProducerConfig%28props%2C+doLog%29+constructor+protected+access

As for establishing the rules going forward, I don't personally have the time to drive that right now but I did file jira ticket for a followup KIP to avoid these things in the future: https://issues.apache.org/jira/browse/KAFKA-15782

Reply:
Btw, what is the use case? If this is for external applications to use, then it does seem like requiring a KIP makes sense.